### PR TITLE
Extend benchmarking module

### PR DESCRIPTION
**Pull request by aaronkl**
This PR augments the benchmarking functionality of emutkit by:
 - passing a complete loop state, rather than just X and Y, to the outer loop during benchmarking 
 - enable the outer loop to have multiple event handler in order to allow to measure multiple metrics
 - adds a new metric that computes the cumulative cost after each function evaluation
 - update the unit test to the new interface

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: emukit/core/loop/outer_loop.py [hunk 1]
```diff
@@ -51,8 +51,8 @@ def __init__(self, candidate_point_calculator: CandidatePointCalculator,
         if self.loop_state is None:
             self.loop_state = LoopState([])
 
-        self.loop_start_event = EventHandler()
-        self.iteration_end_event = EventHandler()
+        self.loop_start_event = []
```
Comment (marpulli) [inline emukit/core/loop/outer_loop.py]: I don't think this is necessary. The `EventHandler` already acts as a list. You can do `self.loop_start_event.append(new_event_fcn)` and doing `self.iteration_end_event(self, self.loop_state)` with call all functions attached to the event. Let me know if I'm missing something though

Comment (apaleyes) [reply emukit/core/loop/outer_loop.py]: That is correct, and is the primary reason EventHandler was introduced: to handle multiple event handlers

**Comment by marpulli** [comment]
This looks good apart from the EventHandler stuff and a unit test for the new CumulativeCost metric

File: emukit/core/loop/outer_loop.py (hunk 1 above)
Comment (aaronkl) [reply emukit/core/loop/outer_loop.py]: ah I see, that makes sense

File: emukit/benchmarking/loop_benchmarking/metrics.py [hunk 2]
```diff
@@ -17,7 +17,8 @@ class MeanSquaredErrorMetric(Metric):
     """
     Mean squared error metric stored in loop state metric dictionary with key "mean_squared_error".
     """
-    def __init__(self, x_test: np.ndarray, y_test: np.ndarray, name: str='mean_squared_error'):
+
+    def __init__(self, x_test: np.ndarray, y_test: np.ndarray, name: str = 'mean_squared_error'):
```
Comment (apaleyes) [inline emukit/benchmarking/loop_benchmarking/metrics.py]: actually the one without spaces is what PEP recommends: "Don't use spaces around the = sign when used to indicate a keyword argument"

applies here and below

File: emukit/benchmarking/loop_benchmarking/metrics.py [hunk 3]
```diff
@@ -81,3 +84,33 @@ def reset(self) -> None:
         Resets the start time
         """
         self.start_time = time()
+
+
+class CumulativeCostMetric(Metric):
+    """
+    The result is stored in the "metrics" dictionary in the loop state with the key "cumulative_costs"
```
Comment (apaleyes) [inline emukit/benchmarking/loop_benchmarking/metrics.py]: Can you extend that a bit, with a description of what is understood as "cumulative cost"?

File: emukit/benchmarking/loop_benchmarking/metrics.py [hunk 4]
```diff
@@ -81,3 +84,33 @@ def reset(self) -> None:
...
+    def __init__(self, name: str = 'cumulative_costs'):
+        self.name = name
+        self.cumulative_costs = np.array([0.0])
+        self.last_observed_iter = 0
+
+    def evaluate(self, loop: OuterLoop, loop_state: LoopState) -> np.ndarray:
+        """
+        Accumulates the cost of each function evaluation
+        :param loop: Outer loop
+        :param loop_state: Object containing history of the loop that we add results to
```
Comment (apaleyes) [inline emukit/benchmarking/loop_benchmarking/metrics.py]: Add "returns" section

File: emukit/benchmarking/loop_benchmarking/benchmarker.py [hunk 5]
```diff
@@ -1,5 +1,6 @@
 import logging
 import numpy as np
+from functools import partial
```
Comment (apaleyes) [inline emukit/benchmarking/loop_benchmarking/benchmarker.py]: Never heard of "partial" before. Thanks, learned something new today!

**Pull request merged**